pkg/reconciler/namespacecontroller/namespace_controller.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package namespacecontroller
+
+import (
+	"context"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"kpt.dev/configsync/pkg/status"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// NamespaceController is a controller that watches for Namespace events and
+// sets a flag to indicate whether the Reconciler thread needs to reconcile for
+// this namespace event.
+type NamespaceController struct {
+	client client.Client
+
+	// eventMux is a RMMutex because the cache is read more often than writen.
+	eventMux sync.RWMutex
+
+	// newNSEvent indicates whether the namespace controller fires up a new namespace event.
+	newNSEvent bool
```

Comment:
Suggest naming this field and the setters/getters after what it's intended to do rather than what caused it.

Like, maybe `namespacesUpdated` or `resyncRequired`.

Reply:
The field is in the namespace controller, but it is the Reconciler that takes the action. If the field is in the reconciler's cache, yes, it should indicate what it is intended to do.
Looking at how it is invoked, it is `nsController.HasNewNSEvent()` in the Reconciler, which makes sense. The reconciler resyncs when the namespace controller has a new event.
If we rename to resyncRequired, then it looks like `nsController.ResyncRequired`, it sounds like the namespace controller requires a new reconciliation, which seems more confusing.

Reply:
Just changes that require resync.

Reply:
SG. Done.